src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs

```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
+                        {
+                            goto case State.AuthorizationShortCircuit;
+                        }
+
+                        goto case State.AuthorizationNext;
+                    }
+
+                case State.AuthorizationShortCircuit:
                     {
-                        // If we get here then the filter didn't call 'next' indicating a short circuit
-                        _resourceExecutedContext = new ResourceExecutedContext(_resourceExecutingContext, _filters)
+                        _logger.AuthorizationFailure((IFilterMetadata)state);
+
+                        // If an authorization filter short circuits, the result is the last thing we execute
+                        // so just return that task instead of calling back into the state machine.
+                        isCompleted = true;
+                        return InvokeResultAsync(_authorizationContext.Result);
+                    }
+
+                case State.AuthorizationEnd:
+                    {
+                        goto case State.ResourceBegin;
+                    }
+
+                case State.ResourceBegin:
+                    {
+                        _cursor.Reset();
+                        goto case State.ResourceNext;
+                    }
+
+                case State.ResourceNext:
+                    {
+                        var current = _cursor.GetNextFilter<IResourceFilter, IAsyncResourceFilter>();
+                        if (current.FilterAsync != null)
                         {
-                            Canceled = true,
-                            Result = _resourceExecutingContext.Result,
-                        };
+                            if (_resourceExecutingContext == null)
+                            {
+                                _resourceExecutingContext = new ResourceExecutingContext(_controllerContext, _filters);
+                            }
+
+                            state = current.FilterAsync;
+                            goto case State.ResourceAsyncBegin;
+                        }
+                        else if (current.Filter != null)
+                        {
+                            if (_resourceExecutingContext == null)
+                            {
+                                _resourceExecutingContext = new ResourceExecutingContext(_controllerContext, _filters);
+                            }
+
+                            state = current.Filter;
+                            goto case State.ResourceSyncBegin;
+                        }
+                        else if (scope == Scope.Resource)
+                        {
+                            Debug.Assert(_resourceExecutingContext != null);
+                            goto case State.ResourceInside;
+                        }
+                        else
+                        {
+                            Debug.Assert(scope == Scope.Invoker);
```

Comment:
Similar comment here, something along the lines of "there were no resource filters to invoke"